src/adapt/proto.ts

```diff
@@ -0,0 +1,478 @@
+  const deps: Record<string, FileDescriptorProto> = {};
+  for (const field of schema.fields || []) {
+    fNumber += 1;
+    const currentScope = `${scope}_${field.name}`;
+    if (field.type === TableFieldSchema.Type.STRUCT) {
+      const subSchema: TableSchema = {
+        fields: field.fields,
+      };
+      const fd = convertStorageSchemaToFileDescriptorInternal(
+        subSchema,
+        currentScope,
+        useProto3
+      );
+      fd.file?.forEach(f => {
+        if (f.name) {
+          deps[f.name] = f;
+        }
+      });
+      const fdp = convertTableFieldSchemaToFieldDescriptorProto(
+        field,
+        fNumber,
+        currentScope,
+        useProto3
+      );
+      fields.push(fdp);
+    } else {
+      const fdp = convertTableFieldSchemaToFieldDescriptorProto(
+        field,
+        fNumber,
+        currentScope,
+        useProto3
+      );
+      fields.push(fdp);
+    }
+  }
+
+  const dp = new DescriptorProto({
+    name: scope,
+    field: fields,
+  });
+
+  const depsNames: string[] = Object.keys(deps);
+  const syntax = useProto3 ? 'proto3' : 'proto2';
+  const fdp = new FileDescriptorProto({
+    messageType: [dp],
+    name: `${scope}.proto`,
+    syntax,
+    dependency: depsNames,
+  });
+
+  const fds = new FileDescriptorSet({
+    file: [fdp, ...Object.values(deps)],
+  });
+
+  return fds;
+}
+
+// normalizeDescriptor builds a self-contained DescriptorProto suitable for communicating schema
+// information with the BigQuery Storage write API.  It's primarily used for cases where users are
+// interested in sending data using a predefined protocol buffer message.
+export function normalizeDescriptor(fds: FileDescriptorSet): DescriptorProto {
+  //let dp = new DescriptorProto();
+  let dp: DescriptorProto | null = null;
+  let fdpName;
+  if (fds.file.length > 0) {
+    // search root descriptor
+    const fdp = fds.file[0];
+    fdpName = fdp.name;
+    if (fdp.messageType && fdp.messageType.length > 0) {
+      dp = new DescriptorProto(fdp.messageType[0]);
+    }
+  }
+  if (!dp) {
+    throw Error('root descriptor not found');
+  }
+  for (const fdp of fds.file) {
+    if (fdp.name === fdpName) {
+      continue;
+    }
+    if (!dp.nestedType) {
+      dp.nestedType = [];
+    }
+    if (!fdp.messageType) {
+      continue;
+    }
+    for (const nestedDP of fdp.messageType) {
+      dp.nestedType.push(
+        normalizeDescriptorProto(new DescriptorProto(nestedDP))
+      );
+    }
+  }
+  return normalizeDescriptorProto(dp);
+}
+
+function normalizeDescriptorProto(dp: DescriptorProto): DescriptorProto {
+  dp.name = normalizeName(dp.name);
+  for (const f of dp.field) {
+    if (f.proto3Optional) {
+      f.proto3Optional = null;
+    }
+    if (f.oneofIndex) {
+      f.oneofIndex = null;
+    }
+  }
+  return dp;
+}
+
+function normalizeName(name: string): string {
+  return name.replace(/\./, '_');
+}
+
+export function fileDescriptorSetToNamespace(
```

Comment:
regarding the use of `Namespace`:  this seems to be describing a descriptor pool.  Is Namespace the term that the node library uses for a descriptor pool?  Or is namespace something other than a pool?

Reply:
`Namespace` in the Node protobuf land is a JSON representation of the DescriptorProto with all dependencies. It's how you can dynamic load structs/classes to be used. For reference: https://protobufjs.github.io/protobuf.js/index.html#using-json-descriptors